dynamic/dynamic_message.go

```diff
@@ -682,7 +682,8 @@ func (m *Message) internalSetField(fd *desc.FieldDescriptor, val interface{}) {
 					}
 				}
 			}
-			if equal {
+
+			if equal && (fd.GetOneOf() == nil) {
```

Comment:
I think we can eliminate the work to test if the value is a default value since it is cheap to test if it is a one-of. So you can move this check above to line 656 Instead, we can more cheaply determine if it's part of a one-of above at line 656:

```go
if m.md.IsProto3() && fd.GetOneOf() == nil {
    // proto3 considers fields that are set to their zero value as unset
    // (except for fields in a oneof)
    if fd.IsRepeated() {
```

Also no need to surround the `==` expression with parentheses.